Documentation/design-docs/WinRT-activation.md

```diff
@@ -0,0 +1,51 @@
+
+In the native (C++/CX, C++/WRL, or C++/WinRT) world, building a WinRT Component named `MyComponent` produces files named `MyComponent.dll`, `MyComponent.winmd`. In this world, the `winmd` contains the metadata describing the types provided by the component, and the code implementing said types is compiled into the `dll`. When an application wants to activate a component, it will call [`RoGetActivationFactory`](https://docs.microsoft.com/windows/desktop/api/roapi/nf-roapi-rogetactivationfactory). The operating system will then search through various areas in the filesystem to find the correct component for the class name. Starting in Windows 10 19H1, there is now support for declaring that specific classes come from components that live side-by-side with the application, similar to Reg-Free COM. After finding the correct component, the OS will load the component via `CoLoadLibrary` and then call the `DllGetActivationFactory` entrypoint, which gets an activation factory for the class by name.
+
+## Proposed Managed WinRT Activation
+
+In the managed (.NET) world, we put all of the code that implements the component into the `winmd`. So, when running in an AppContainer/`.appx`, a .NET component only needs one file. However, when outside of an AppContainer, there needs to be some sort of host to activate runtime. We will supply a host that implements the [`DllGetActivationFactory`](https://docs.microsoft.com/previous-versions//br205771(v=vs.85)) entrypoint. When `DllGetActivationFactory` is called, the following will occur:
+
+1) If a [`.runtimeconfig.json`](https://github.com/dotnet/cli/blob/master/Documentation/specs/runtime-configuration-file.md) file exists adjacent to the shim assembly (`<shim_name>.runtimeconfig.json`), that file will be used to describe CLR configuration details. The documentation for the `.runtimeconfig.json` format defines under what circumstances this file may be optional.
+2) Using the existing `hostfxr` library, attempt to discover the desired CLR and target [framework](https://docs.microsoft.com/en-us/dotnet/core/packages#frameworks).
+   * If a CLR is active with the process, the requested CLR version will be validated against that CLR. If version satisfiability fails, activation will fail.
+   * If a CLR is **not** active with the process, an attempt will be made to create a satisfying CLR instance.
+   * Failure to create an instance will result in activation failure.
+3) A request to the CLR will be made to load the assembly from memory and get the entry-point.
+   * The ability to load an assembly from memory will require exposing a new function that can be called from `hostfxr`, as well as a new API in `System.Private.CoreLib` on a new class in `Internal.Runtime.InteropServices`:
+
+```csharp
+namespace Internal.Runtime.InteropServices.WindowsRuntime
+{
+    public static class ActivationFactoryLoader
+    {
+        public static int GetActivationFactory(
+            IntPtr componentPath,
```

Comment:
Since this is a string, can it be declared as one, or do we really have to do manual marshaling...?

Reply:
The marshalling infrastructure gets confused on marshalling strings when using the "hosting" marshalling model (where there is no way to specify a CharSet). I'll change it this parameter to be a `char*` to better signify what it is supposed to be.